services/cd-service/pkg/service/batch_test.go

```diff
@@ -0,0 +1,282 @@
+/*This file is part of kuberpult.
+
+Kuberpult is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Kuberpult is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.
+
+Copyright 2021 freiheit.com*/
+package service
+
+import (
+	"context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"os/exec"
+	"path"
+	"testing"
+
+	"github.com/freiheit-com/kuberpult/pkg/api"
+	"github.com/freiheit-com/kuberpult/services/cd-service/pkg/config"
+	"github.com/freiheit-com/kuberpult/services/cd-service/pkg/repository"
+)
+
+func TestBatchService(t *testing.T) {
+	tcs := []struct {
+		Name  string
+		Setup []repository.Transformer
+		Test  func(t *testing.T, svc *BatchServer)
+	}{
+		{
+			Name: "5 sample actions",
+			Setup: []repository.Transformer{
+				&repository.CreateEnvironment{
+					Environment: "production",
+					Config:      config.EnvironmentConfig{Upstream: &config.EnvironmentConfigUpstream{Latest: true}},
+				},
+				&repository.CreateApplicationVersion{
+					Application: "test",
+					Manifests: map[string]string{
+						"production": "manifest",
+					},
+				},
+				&repository.CreateEnvironmentLock{   // will be deleted
+					Environment: "production",
+					LockId: "1234",
+					Message: "EnvLock",
+				},
+				&repository.CreateEnvironmentApplicationLock{  // will be deleted
+					Environment: "production",
+					Application: "test",
+					LockId: "5678",
+					Message: "AppLock",
+				},
+			},
+			Test: func(t *testing.T, svc *BatchServer) {
+				opDeploy := &api.BatchAction_Deploy{
```

Comment:
in general: try to keep the `Test` function small.
here specificially: The array of `BatchAction`s belongs in the `tcs` struct.
If it gets more complicated, like in the 2nd test, the BatchActions can also be returned from a function (e.g. `createNBatchActions`), but they should be part of the tcs struct.